spec/unit/readertoc_spec.lua

```diff
@@ -1,19 +1,23 @@
 describe("Readertoc module", function()
-    local DocumentRegistry, ReaderUI, DEBUG
+    local DocumentRegistry, Event, ReaderUI, DEBUG
     local readerui, toc, toc_max_depth, title
 
     setup(function()
         require("commonrequire")
         DocumentRegistry = require("document/documentregistry")
+        Event = require("ui/event")
         ReaderUI = require("apps/reader/readerui")
         DEBUG = require("dbg")
 
         local sample_epub = "spec/front/unit/data/juliet.epub"
+
         readerui = ReaderUI:new{
             document = DocumentRegistry:openDocument(sample_epub),
         }
         -- reset book to first page
         readerui.rolling:onGotoPage(0)
+        readerui.document:setSpaceCondensing(75)
```

Comment:
@poire-z Could this be related to the cache file btw? It's not quite clear to me why this setting pollutes.

Reply:
I'm not sure, but I don't think so.
It's possible the default is 50 in crengine if not specified.
But we should set it in `ReaderFont:onReadSettings()`:
https://github.com/koreader/koreader/blob/27ddd6f313cf4cc609f6ec064e09ae0f0a049beb/frontend/apps/reader/modules/readerfont.lua#L156-L158

and this setting is accounted in the HASH that is used to see if the cached rendering info is still valid, or if a re-rendering is to be done:
https://github.com/koreader/crengine/blob/cf8a533b20cf8be30b30e79aa5ff0a0ed8b7dc61/crengine/src/lvtinydom.cpp#L10147-L10153

So, any change of minSpaceCondensing should be noticed and cache reused or not accordingly.

Reply:
I may have very well misidentified space condensing as the culprit. A `printf("_minSpaceCondensingPercent: %i\n", _minSpaceCondensingPercent);` shows it to be at 75. Getting rid of the 50 in readerrolling or setting it to 75 in readertoc seemed to solve the problem.

I can in any case confirm that the cache is at fault, tested by removing it manually. I've attached the cache generated by readerrolling_spec (breaks-test) and by readertoc_spec (works-test) in case it might help.
[cr3cache.zip](https://github.com/koreader/koreader/files/3119911/cr3cache.zip)

```sh
$ md5sum juliet.epub.aaa10123.1.cr3.works-test 
d70ea3f15cc9edd83a0ed4442f5d8c91  juliet.epub.aaa10123.1.cr3.works-test
$ md5sum juliet.epub.aaa10123.1.cr3.breaks-test 
88df0fbee32b3b9aa9225c5610c57da6  juliet.epub.aaa10123.1.cr3.breaks-test
```

So for the purposes of the test it might be best to remove the cache in teardown, even if it'd result in slightly longer test runs. On the flipside, it might hint at some bug. I'm not quite sure. :-/

Reply:
I'm not sure a md5sum on the cache file is a hint it has structurally changed, although I don't see/remember what may change a bit in it in different invocations of KOreader with no setting change:
```
$ md5sum Saint-Empire-ique.FR.epub.e32923e8.1.cr3 ; rm Saint-Empire-ique.FR.epub.e32923e8.1.cr3
    017dba090de3fcee9c52ea032a99b6d8  Saint-Empire-ique.FR.epub.e32923e8.1.cr3
$ md5sum Saint-Empire-ique.FR.epub.e32923e8.1.cr3 ; rm Saint-Empire-ique.FR.epub.e32923e8.1.cr3
    ce0bc5f6bcdd6d7bb402b1ab31190e34  Saint-Empire-ique.FR.epub.e32923e8.1.cr3
$ md5sum Saint-Empire-ique.FR.epub.e32923e8.1.cr3 ; rm Saint-Empire-ique.FR.epub.e32923e8.1.cr3
    be6e9e7648294d6d1415bdc20f3cc22c  Saint-Empire-ique.FR.epub.e32923e8.1.cr3
```

Reply:
I confirmed the behavior. The md5sum merely illustrates that there seems to be a subtle difference between the generated cache even though it has exactly the same size.